toxcore/forwarding.c

```diff
@@ -0,0 +1,377 @@
+/* SPDX-License-Identifier: GPL-3.0-or-later
+ * Copyright © 2019-2022 The TokTok team.
+ */
+
+#include "forwarding.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "DHT.h"
+#include "ccompat.h"
+#include "timed_auth.h"
+
+struct Forwarding {
+    const Logger *log;
+    DHT *dht;
+    Mono_Time *mono_time;
+    Networking_Core *net;
+
+    uint8_t hmac_key[CRYPTO_HMAC_KEY_SIZE];
+
+    forward_reply_cb *forward_reply_callback;
+    void *forward_reply_callback_object;
+
+    forwarded_request_cb *forwarded_request_callback;
+    void *forwarded_request_callback_object;
+
+    forwarded_response_cb *forwarded_response_callback;
+    void *forwarded_response_callback_object;
+};
+
+DHT *forwarding_get_dht(Forwarding *forwarding)
+{
+    return forwarding->dht;
+}
+
+#define SENDBACK_TIMEOUT 3600
+
+bool send_forward_request(Networking_Core *net, const IP_Port *forwarder,
+                          const uint8_t *chain_keys, uint16_t chain_length,
+                          const uint8_t *data, uint16_t data_length)
+{
+    const uint16_t len = forward_chain_packet_size(chain_length, data_length);
+    VLA(uint8_t, packet, len);
+
+    return create_forward_chain_packet(chain_keys, chain_length, data, data_length, packet)
+           && sendpacket(net, forwarder, packet, len) == len;
+}
+
+uint16_t forward_chain_packet_size(uint16_t chain_length, uint16_t data_length)
+{
+    return chain_length * (1 + CRYPTO_PUBLIC_KEY_SIZE) + data_length;
+}
+
+bool create_forward_chain_packet(const uint8_t *chain_keys, uint16_t chain_length,
+                                 const uint8_t *data, uint16_t data_length,
+                                 uint8_t *packet)
+{
+    if (chain_length == 0 || chain_length > MAX_FORWARD_CHAIN_LENGTH
+            || data_length > MAX_FORWARD_DATA_SIZE) {
+        return false;
+    }
+
+    uint16_t i = 0;
```

Comment:
`i` could have a better name. It's not a loop iterator, but rather a byte offset in the packet.

Reply:
made it `offset`